### PR TITLE
[versioning][android] rewrite expoview versioning to ts

### PR DESCRIPTION
**Pull request by wkozyra95**
# Why

getting it on par with iOS versioning

# How

- reuse existing abstractions from iOS versioning
- mostly keep the same logic as the original scripts(I'll add more info in the comments)
- I left some parts that are commented out, I plan to remove them, just left them for the review

# Test Plan

- comment out everything except expoview copy step in https://github.com/expo/expo/blob/main/tools/src/versioning/android/index.ts#L633
- Run `et add-sdk --platform android --sdkVersion 46.0.0`
- commit changes
- switch back to old shell script
- delete versioned code and run add-sdk again
- verify no changes

Note: It was tested on linux

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: tools/src/Transforms.ts [hunk 1]
```diff
@@ -17,11 +23,15 @@ export function transformString(
...
     return input;
   }
-  return transforms.reduce(
-    // @ts-ignore @tsapeta: TS gets crazy on `replaceWith` being a function.
-    (acc, { find, replaceWith }) => acc.replace(find, replaceWith),
-    input
-  );
+  return transforms.reduce((acc, transform) => {
+    if ('fn' in transform) {
+      return (transform as RawTransform).fn(acc);
```
Comment (tsapeta) [inline tools/src/Transforms.ts:28]:
```suggestion
    if ('transform' in transform) {
      return (transform as RawTransform).transform(acc);
```

File: tools/src/Transforms.types.ts [hunk 2]
```diff
@@ -13,6 +13,12 @@ export type StringTransform = {
   replaceWith: string | ((substring: string, ...args: any[]) => string);
 };
 
+export type RawTransform = {
+  fn: (text: string) => string;
```
Comment (tsapeta) [inline tools/src/Transforms.types.ts:17]:
```suggestion
  transform: (text: string) => string;
```

File: tools/src/versioning/android/copyExpoview.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    return !shouldDelete;
+  });
+  return filteredLines.join('\n');
+}
+
+export async function copyExpoviewAsync(sdkVersion: string, androidDir: string): Promise<void> {
+  const abiVersion = `abi${sdkVersion.replace(/\./g, '_')}`;
+  const targetDirectory = path.join(androidDir, `versioned-abis/expoview-${abiVersion}`);
+  const sourceDirectory = path.join(androidDir, 'expoview');
+  const transforms: FileTransforms = {
```
Comment (tsapeta) [inline tools/src/versioning/android/copyExpoview.ts:35]: Could you move it to a separate file?

File: tools/src/versioning/android/copyExpoview.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        replaceWith: `import $1${abiVersion}.${pkg}`,
+      })),
+      //This approach would be safer, but I don't think it's necessary
+      //...packagesToKeep.map((pkg: string) => ({
+      //  paths: './src/main/java/**/*.{java,kt}',
+      //  find: new RegExp(`temporarydonotversion\\\.${escapeRegExp(pkg)}`, 'g'),
+      //  replaceWith: pkg,
+      //})),
+      {
+        paths: [`./src/main/java/**/*.{java,kt}`],
```
Comment (tsapeta) [inline tools/src/versioning/android/copyExpoview.ts:113]:
```suggestion
        paths: `./src/main/java/**/*.{java,kt}`,
```

**Review by tsapeta — APPROVED** [review]

**Review by Kudo — APPROVED** [review]
looked all the comments and all make sense to me.
there's just a nit for the `fn` transform. leaving it and see how do you think about it.

**Review by tsapeta — APPROVED** [review]

**Pull request merged**